onboard/catkin_ws/src/acoustics/scripts/acoustics_wrapper.py

```diff
@@ -76,19 +94,31 @@ def get_processing_angle(self, wrapper_goal, processing_files, guess):
         return self.client_processing.get_result()
 
     def execute(self, wrapper_goal):
+
         time_now = datetime.now()
         guess_export_name = "date_" + time_now.strftime("%m_%d_%Y_%H_%M_%S") + "_guess"
         processing_export_name = "date_" + time_now.strftime("%m_%d_%Y_%H_%M_%S") + "_processing"
 
-        self.publish_feedback(0, "Starting sampling for guess")
-        guess_sampling = self.saleae_sampling(1, guess_export_name)
-        self.publish_feedback(1, "Sampling for guess complete, starting guess")
+        if wrapper_goal.gen_data == True:
+            self.publish_feedback(0, "Generating data for guess")
+            guess_sampling = self.generate_data(1, wrapper_goal)
+            self.publish_feedback(1, "Generating data complete, starting guess")
+
+            guess = self.get_guess(wrapper_goal, guess_sampling.file_paths)
+            self.publish_feedback(2, "Guess complete, generating data for processing")
+
+            processing_sampling = self.saleae_sampling(2, processing_export_name)
+            self.publish_feedback(3, "Generating data for processing complete, starting processing")
+        else:
+            self.publish_feedback(0, "Starting sampling for guess")
+            guess_sampling = self.saleae_sampling(1, guess_export_name)
+            self.publish_feedback(1, "Sampling for guess complete, starting guess")
 
-        guess = self.get_guess(wrapper_goal, guess_sampling.file_paths)
-        self.publish_feedback(2, "Guess complete, starting sampling for processing")
+            guess = self.get_guess(wrapper_goal, guess_sampling.file_paths)
+            self.publish_feedback(2, "Guess complete, starting sampling for processing")
 
-        processing_sampling = self.saleae_sampling(2, processing_export_name)
-        self.publish_feedback(3, "Sampling for processing complete, starting processing")
+            processing_sampling = self.saleae_sampling(2, processing_export_name)
+            self.publish_feedback(3, "Sampling for processing complete, starting processing")
```

Comment:
This needs to be redesigned, you have code duplication here in guessing and processing occurring in the if statements. You should pass in the fn you want to call as a argument into this execute function, and when you create the action server, assign it there using a lambda to forward the argument based on what type of server you're creating, which was based on that param you read in.